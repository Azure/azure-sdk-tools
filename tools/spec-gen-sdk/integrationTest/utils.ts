import * as fs from 'fs-extra';
import _rimraf from 'rimraf';
import { Octokit } from '@octokit/rest';
import { sdkAutomationCliConfig } from '../src/cli/config';
import { getAuthenticatedOctokit } from '../src/utils/githubUtils';
import { deleteFolder } from '@ts-common/azure-js-dev-tools';
import * as winston from 'winston';
import path from 'path';
import simpleGit, { ResetMode, SimpleGit, SimpleGitOptions } from 'simple-git';
import { getCurrentBranch, gitAddAll, gitGetCommitter, gitGetDiffFileList, gitSetRemoteWithAuth } from '../src/utils/gitUtils';
import { sdkAutoLogLevels } from '../src/automation/logging';
import { simpleGitOptions } from '../src/automation/workflow';

const fixturePrefix = 'integrationTest/fixtures';
export const workPrefix = sdkAutomationCliConfig.workingFolder;
export const repoOwner = sdkAutomationCliConfig.specRepo.split('/')[0];

let runId = sdkAutomationCliConfig.testRunId;
if (!runId) {
  runId = Math.random().toString(36).substring(2, 8);
}
export const getRunIdPrefix = () => `test-${runId}`;
const logger = winston.createLogger({
  levels: sdkAutoLogLevels.levels,
  transports: [new winston.transports.Console({ level: 'error' })]
});
const [githubClient, getGithubAccessToken] = getAuthenticatedOctokit(
  {
    token: sdkAutomationCliConfig.githubToken,
    id: sdkAutomationCliConfig.githubApp.id,
    privateKey: sdkAutomationCliConfig.githubApp.privateKey
  },
  logger
);

export const getTestGithubClient = async (): Promise<Octokit> => {
  return githubClient;
};

export const getTestGitClient = async (repoName: string) => {
  const repoFolder = path.join(workPrefix, repoName);
  fs.mkdirpSync(repoFolder);

  const repo = simpleGit({ ...simpleGitOptions, baseDir: repoFolder });
  await repo.init(false);
  return repo;
};

export const cleanTestRepos = async (repos: string[]) => {
  const github = await getTestGithubClient();

  for (const repoName of repos) {
    console.log(`Cleaning up ${repoOwner}/${repoName}`);
    await deleteFolder(`${workPrefix}/${repoName}`);
    try {
      await github.repos.delete({ owner: repoOwner, repo: repoName });
    } catch (e) {
      console.log(`Failed to delete ${repoOwner}/${repoName}: ${e.message} ${e.stack}`);
    }
  }
};

type ReplaceFileRecord = {
  [filePath: string]: {
    search: string;
    replace: string;
  }[];
};
export const applyReplaceFileRecord = (workPath: string, toReplace?: ReplaceFileRecord) => {
  if (toReplace !== undefined) {
    for (const [filePath, replaceArr] of Object.entries(toReplace)) {
      let content = fs.readFileSync(`${workPath}/${filePath}`).toString();
      for (const { search, replace } of replaceArr) {
        content = content.replace(search, replace);
      }
      fs.writeFileSync(`${workPath}/${filePath}`, content);
    }
  }
};

export const initializeGithubRepoFromLocalFixture = async (
  fixtureName: string,
  repoName: string,
  toReplace?: ReplaceFileRecord,
  baseBranch: string = 'master'
) => {
  console.log(`Initializing https://github.com/${repoOwner}/${repoName}`);
  const fixturePath = path.join(fixturePrefix, fixtureName);
  const workPath = path.join(workPrefix, repoName);
  const github = await getTestGithubClient();

  // Copy file
  await fs.remove(workPath);
  await fs.mkdirp(workPath);
  await fs.copy(fixturePath, workPath);
  applyReplaceFileRecord(workPath, toReplace);
  const repo: SimpleGit = await getTestGitClient(repoName);

  // Create repo if not exist
  try {
    await github.repos.delete({ owner: repoOwner, repo: repoName });
  } catch (e) {
    // Repo not found. Pass
  }
  await github.repos.createInOrg({ name: repoName, org: repoOwner });
  try {
    await github.repos.get({ owner: repoOwner, repo: repoName });
  } catch (e) {}

  // Checkout branch
  await gitGetCommitter(repo);
  await repo.add([]).commit(`Init repo`)
  const commit = await repo.revparse('HEAD');
  const currentBranch = await getCurrentBranch(repo);
  if (baseBranch !== currentBranch) {
    await repo.raw(['branch', '--copy', commit, baseBranch, '--force']);
    await repo.checkoutLocalBranch(baseBranch);
  }

  // Add and commit
  await gitAddAll(repo);
  const diff = await repo.diff(['--name-status', 'HEAD']);
  let fileList = await gitGetDiffFileList(diff);
  await repo.add(fileList).commit(`Init repo with fixture ${fixtureName}`)

  // Push
  await gitSetRemoteWithAuth({ logger, getGithubAccessToken }, repo, 'origin', { owner: repoOwner, name: repoName });
  await repo.push([
    'origin',
    `+refs/heads/${baseBranch}:refs/heads/${baseBranch}`
  ]);
  console.log(`Initialized https://github.com/${repoOwner}/${repoName}`);
};

export const createPRWithPatch = async (
  repoName: string,
  patchName: string,
  branchName: string,
  options: {
    toReplace?: ReplaceFileRecord;
    baseBranch?: string;
    fetchBase?: boolean;
  } = {}
) => {
  const baseBranch = options.baseBranch ?? 'master';

  console.log(`Creating PR in '${repoOwner}/${repoName}' with patch '${patchName}' on branch '${branchName}'`);
  const workPath = `${workPrefix}/${repoName}`;
  const patchPath = `${fixturePrefix}/${patchName}`;
  const github = await getTestGithubClient();
  const repo = await getTestGitClient(repoName);

  await gitAddAll(repo);
  const header = await repo.revparse('HEAD')
  await repo.reset(ResetMode.HARD, [header]);
  await repo.checkout(header, ['--force']);

  if (options.fetchBase) {
    await repo.fetch([
       'origin',
       '--update-head-ok',
       '--force',
       `+refs/heads/${baseBranch}:refs/remotes/origin/${baseBranch}`
     ]);
    await repo.mergeFromTo(`origin/${baseBranch}`, baseBranch);
  }

  await repo.checkout(branchName, ['--force']);
  const baseCommit = await repo.revparse('HEAD');
  await repo.raw(['branch', '--copy', baseCommit, baseBranch, '--force']);
  await repo.checkout(branchName, ['--force']);

  await fs.copy(patchPath, workPath);
  applyReplaceFileRecord(workPath, options.toReplace);

  await gitAddAll(repo);
  const diff = await repo.diff(['--name-status', 'HEAD']);
  let fileList = await gitGetDiffFileList(diff);
  await gitGetCommitter(repo);
  await repo.add(fileList).commit(`Apply patch ${patchName}`)

  // Push
  await gitSetRemoteWithAuth({ logger, getGithubAccessToken }, repo, 'origin', { owner: repoOwner, name: repoName });
  await repo.push([
    'origin',
    `+refs/heads/${branchName}:refs/heads/${branchName}`
  ]);

  let pr = await github.pulls.create({
    owner: repoOwner,
    repo: repoName,
    head: branchName,
    base: baseBranch,
    title: branchName
  });
  console.log(pr.data.html_url);

  // Wait for mergable data
  while (pr.data.mergeable === null) {
    pr = await github.pulls.get({ owner: repoOwner, repo: repoName, pull_number: pr.data.number });
  }

  return pr.data.number;
};

export const getPRListInfo = async (repoName: string) => {
  const github = await getTestGithubClient();
  // Wait for 5 seconds to make sure we could get the latest pull request list
  await new Promise((resolve) => setTimeout(resolve, 5000));
  const { data: sdkPRs } = await github.pulls.list({
    owner: repoOwner,
    repo: repoName,
    state: 'all'
  });
  const result = sdkPRs.map((pr) => ({
    number: pr.number,
    state: pr.state,
    title: pr.title,
    base: pr.base.ref,
    head: pr.head.ref
  }));
  result.sort((a, b) => a.number - b.number);
  return result;
};

export const getLoggerTextForAssert = async (
  _logger: { allLogs: string[] },
  specRepoName: string,
  specPRNumber: number,
  sdkRepoName: string
) => {
  let text = _logger.allLogs.join('\n');
  text = text.replace(new RegExp(fs.realpathSync('.'), 'g'), '.');
  text = text.replace(new RegExp(repoOwner, 'g'), '<org>');
  text = text.replace(new RegExp(specRepoName, 'g'), '<spec_repo>');
  text = text.replace(new RegExp(sdkRepoName, 'g'), '<sdk_repo>');

  text = text.replace(/ [a-z0-9]{40}/g, ` <commit_sha>`);

  const version = require('pkginfo').read(module).package.version;
  text = text.replace(new RegExp(version.replace('.', '\\.'), 'g'), '<version>');

  text = text.replace(new RegExp(path.resolve(workPrefix), 'g'), '<work_dir>');

  return text;
};

export const launchTestSDKAutomation = async (automationPromise: Promise<void>, info: string) => {
  console.log(`Launching SDK Automation for ${info}`);
  try {
    await automationPromise;
  } catch (e) {
    console.log(e);
    // Need log for diagnostic so do not fail on exception
  }
};
