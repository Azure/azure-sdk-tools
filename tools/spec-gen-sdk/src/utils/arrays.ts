/**
 * Get whether or not the provided array contains any values.
 * @param values The array to check.
 * @returns Whether or not the provided array contains any values.
 */
export function any<T>(values: T[] | undefined): values is T[] {
    return !!(values && values.length > 0);
}
  

/**
 * Get all of the values within the provided array of values that match the provided condition.
 * @param values The array of values to filter.
 * @param condition The condition to look for within the array of values.
 * @returns The array of values from the original values that match the provided condition.
 */
export function where<T>(values: T[], condition: (value: T) => boolean): T[] {
  const result: T[] = [];
  for (const value of values) {
    if (condition(value)) {
      result.push(value);
    }
  }
  return result;
}

/**
 * Ensure that a value that is either a single value or an array is an array by wrapping single
 * values in an array.
 * @param value The value to ensure is an array.
 * @param conversion The function that will be used to convert the non-array value to an array. This
 * defaults to just creating a new array with the single value.
 * @returns The array value.
 */
export function toArray<T>(value: T | T[], conversion: (valueToConvert: T) => T[] = (valueToConvert: T) => [valueToConvert]): T[] {
  return value instanceof Array ? value : conversion(value);
}
  