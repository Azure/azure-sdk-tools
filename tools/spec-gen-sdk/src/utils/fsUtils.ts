import path from 'path';
import fs from 'fs';
import { WorkflowContext } from '../automation/workflow';
import { SimpleGit } from 'simple-git';
import { TreeType, gitTreeResultToStringArray } from './gitUtils';
import { sdkAutomationCliConfig } from '../cli/config';

export type FsSearchOptions = {
  searchFileRegex: RegExp
} & ({
  rootFolder: string;
} | {
  repo: SimpleGit;
  treeId: string;
  specFolder: string,
});

export const searchRelatedFolder = async (
  filePath: string,
  opts: FsSearchOptions
) => {
  let searchPath = filePath;

  while (searchPath !== '.') {
    const fileName = path.basename(searchPath);
    if (opts.searchFileRegex.exec(fileName)) {
      return searchPath;
    }
    searchPath = path.dirname(searchPath);
  }
  return undefined;
};

export const searchSharedLibrary = async (
  fileList: string[],
  opts: FsSearchOptions
) => {
  const result: { [relatedFolder: string]: string[] } = {};
  fileList.sort();
  let lastFolder: string | undefined = undefined;

  for (const filePath of fileList) {
    if (lastFolder !== undefined && filePath.startsWith(lastFolder)) {
      result[lastFolder].push(filePath);
    }
    const relatedFolder = await searchRelatedFolder(filePath, opts);
    if (relatedFolder === undefined) {
      continue;
    }
    if (result[relatedFolder] === undefined) {
      result[relatedFolder] = [];
    }
    result[relatedFolder].push(filePath);
    lastFolder = relatedFolder;
  }

  return result;
};

export const searchRelatedTypeSpecProjectBySharedLibrary = async (
  sharedLibraries: { [relatedFolder: string]: string[] },
  opts: FsSearchOptions
) => {
  const result: { [relatedFolder: string]: string[] } = {};
  for (const sharedLibrary of Object.keys(sharedLibraries)) {
    const parentFolder = path.dirname(sharedLibrary);
    const fileNames = await getFilesInFolder(parentFolder, opts);
    for (const fileName of fileNames) {
      const filePath = path.join(parentFolder, fileName);
      const subFileNames = await getFilesInFolder(filePath, opts);
      for (const subFileName of subFileNames) {
        if (opts.searchFileRegex.exec(subFileName)) {
          if (!result[filePath]) {
            result[filePath] = [];
          }
          result[filePath] = result[filePath].concat(sharedLibraries[sharedLibrary]);
        }
      }
    }
  }
  return result;
};

export const searchRelatedParentFolders = async (
  fileList: string[],
  opts: FsSearchOptions
) => {
  const result: { [relatedFolder: string]: string[] } = {};
  fileList.sort();
  
  for (const filePath of fileList) {
    const relatedParentFolder = await searchRelatedParentFolder(filePath, opts);
    if (relatedParentFolder === undefined) {
      continue;
    }
    if (result[relatedParentFolder] === undefined) {
      result[relatedParentFolder] = [];
    }
    result[relatedParentFolder].push(filePath);
  }

  return result;
};

export const searchRelatedParentFolder = async (
  filePath: string,
  opts: FsSearchOptions
) => {
  let searchPath = filePath;

  while (searchPath !== '.') {
    const fileNames = await getFilesInFolder(searchPath, opts);
    for (const fileName of fileNames) {
      if (opts.searchFileRegex.exec(fileName)) {
        return searchPath;
      }
    }
    searchPath = path.dirname(searchPath);
  }

  return undefined;
};

const getFilesInFolder = async (searchPath: string, opts: FsSearchOptions): Promise<string[]> => {
  if ('rootFolder' in opts) {
    const p = path.join(opts.rootFolder, searchPath);
    if (!fs.existsSync(p) || !fs.lstatSync(p).isDirectory()) {
      return [];
    }
    return fs.readdirSync(p);
  }

  const workingFolder = sdkAutomationCliConfig.workingFolder;
  const workPath = path.resolve(process.cwd(), workingFolder, opts.specFolder, searchPath);
  const tree = await opts.repo.raw(['ls-tree', `${opts.treeId}`, workPath]);
  const subTree = gitTreeResultToStringArray(tree);
  if (subTree.length === 0 || (subTree.length > 0 && subTree[0].type !== TreeType.TREE)) {
    return [];
  }
  const entryPath = path.join(workPath, '/');
  const treeEntry = await opts.repo.raw(['ls-tree', `${opts.treeId}`, entryPath]);
  const subTreeEntry = gitTreeResultToStringArray(treeEntry);
  return subTreeEntry.map(item => item.file.slice(searchPath.length + 1));
};

export const writeTmpJsonFile = (context: WorkflowContext, fileName: string, content: unknown) => {
  const filePath = path.join(context.tmpFolder, fileName);
  const contentString = JSON.stringify(content, undefined, 2);
  context.logger.info(`Write temp file ${filePath} with content:`);
  context.logger.info(JSON.stringify(content, undefined, 2));
  fs.writeFileSync(filePath, contentString);
};

export const readTmpJsonFile = (context: WorkflowContext, fileName: string): unknown | undefined => {
  const filePath = path.join(context.tmpFolder, fileName);

  if (!fs.existsSync(filePath)) {
    context.logger.warn(`Warning: File ${filePath} not found to read. Please re-run the pipeline if the error is transitient error or report this issue through https://aka.ms/azsdk/support/specreview-channel.`);
    return undefined;
  }

  try {
    context.logger.info(`Read temp file ${filePath} with content:`);
    const contentString = fs.readFileSync(filePath).toString();
    const content = JSON.parse(contentString);
    context.logger.info(JSON.stringify(content, undefined, 2));
    return content;
  } catch (e) {
    context.logger.error(`IOError: Failed to read ${fileName}: ${e.message}. Please re-run the pipeline if the error is retryable or report this issue through https://aka.ms/azsdk/support/specreview-channel.`);
    return undefined;
  }
};

export const deleteTmpJsonFile = (context: WorkflowContext, fileName: string) => {
  const filePath = path.join(context.tmpFolder, fileName);
  if (fs.existsSync(filePath)) {
    fs.unlinkSync(filePath);
  }
};
