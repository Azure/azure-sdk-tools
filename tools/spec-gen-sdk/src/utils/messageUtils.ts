/**
 * Message prefixes for telemetry in the spec-gen-sdk tool
 */
export enum MessagePrefix {
    ToolError = "SGS-ERR",     // Errors from spec-gen-sdk tool
    ExternalError = "EXT-ERR", // Errors from external scripts
    ConfigError = "CONFIG-ERR", // Errors from configuration
    ToolWarning = "SGS-WARN",  // Warnings from spec-gen-sdk tool
    ExternalWarning = "EXT-WARN", // Warnings from external scripts
    ConfigWarning = "CONFIG-WARN", // Warnings from configuration
  }
  
  /**
   * Formats a message with a consistent prefix for telemetry
   * 
   * @param prefix The prefix to use
   * @param message The message
   * @param details Optional additional details
   * @returns Formatted message with prefix
   */
  export function formatMessage(
    prefix: MessagePrefix,
    message: string,
    details?: string
  ): string {
    const prefixStr = `[${prefix}]`;
  
    if (details) {
        return `${prefixStr} ${message}. ${details}`;
    }
  
    return `${prefixStr} ${message}`;
  }
  
  /**
   * Helper for tool errors
   */
  export function toolError(message: string, details?: string): string {
    return formatMessage(MessagePrefix.ToolError, message, details);
  }
  
  /**
   * Helper for external errors
   */
  export function externalError(message: string, details?: string): string {
    return formatMessage(MessagePrefix.ExternalError, message, details);
  }
  
  /**
   * Helper for configuration errors
   */
  export function configError(message: string, details?: string): string {
    return formatMessage(MessagePrefix.ConfigError, message, details);
  }

  /**
   * Helper for tool warnings
   */
  export function toolWarning(message: string, details?: string): string {
  return formatMessage(MessagePrefix.ToolWarning, message, details);
  }
  
  /**
   * Helper for external warnings
   */
  export function externalWarning(message: string, details?: string): string {
    return formatMessage(MessagePrefix.ExternalWarning, message, details);
  }

  /**
   * Helper for configuration warnings
   */
  export function configWarning(message: string, details?: string): string {
    return formatMessage(MessagePrefix.ConfigWarning, message, details);
  }
  
  /**
   * Check if a message already has a prefix
   */
  export function hasPrefix(message: string): boolean {
    return /^\[(SGS|EXT|CONFIG)-(ERR|WARN)\]/.test(message);
  }
