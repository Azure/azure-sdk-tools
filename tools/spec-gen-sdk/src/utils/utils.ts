import _ from "lodash";
import YAML from "yaml";
import { getTypeSpecOutputFolder } from "./typespecUtils";
import { WorkflowContext } from "../automation/workflow";

export function diffStringArrays(left: string[], right: string[]): { diffResult: string[], hasDiff: boolean} {
  const unchanged = right.filter(item => left.includes(item));
  const added = right.filter(item => !left.includes(item));
  const diffResult: string[] = [];

  unchanged.forEach(item => {
    diffResult.push(`\t${item}`);
  });

  added.forEach(item => {
    diffResult.push(`+\t${item}`);
  });

  return {
    diffResult,
    hasDiff: added.length > 0
  };
  
}

export function getValueByKey<T>(
    arrayOfObjects: { [key: string]: T }[],
    targetKey: string
): T | undefined {
    const foundObject = _.find(arrayOfObjects, targetKey);
    return foundObject ? foundObject[targetKey] : undefined;
}

export function removeAnsiEscapeCodes<T extends string | string[]>(messages: T): T {
    // eslint-disable-next-line no-control-regex
    const ansiEscapeCodeRegex = /\x1b\[(\d{1,2}(;\d{0,2})*)?[A-HJKSTfimnsu]/g;
    if (typeof messages === "string") {
        return messages.replace(ansiEscapeCodeRegex, "") as T;
    }
    return messages.map((item) => item.replace(ansiEscapeCodeRegex, "")) as T;
}

export function extractServiceName(path: string): string {
    const match = path.match(/specification\/([^/]*)\//);
    return match ? match[1] : "";
}

// Flag of the readme.md under root of 'resource-manager' or 'data-plane'
const IsReadmeUnderRoot = /specification\/([^/]*)\/([^/]*)\/readme\.md/g;

export function removeDuplicatesFromRelatedFiles(
    relatedTypeSpecProjectFolder: string[] | undefined,
    relatedReadmeMdFiles: string[] | undefined,
    context: WorkflowContext
): string[] {
    const _relatedTypeSpecProjectFolder = relatedTypeSpecProjectFolder || [];
    const _relatedReadmeMdFiles = relatedReadmeMdFiles || [];
    const filteredReadmeMdFiles = _relatedReadmeMdFiles.filter((readmeFile) => {
        const readmeServiceName = extractServiceName(readmeFile);
        const isResourceManager = readmeFile.includes("/resource-manager/");
        const isDataPlane = readmeFile.includes("/data-plane/");
        let typespecOutputFolderRegExp: RegExp;

        return !_relatedTypeSpecProjectFolder.some((typespecFile) => {
            const typespecServiceName = extractServiceName(typespecFile);
            const isManagement = typespecFile.endsWith(".Management");

            if (readmeServiceName === typespecServiceName) {
                if (isManagement && isResourceManager) {
                    context.logger.info(
                        `Exclude the readmeFile from SDK generation: ${readmeFile} because it is duplicated with typespec: ${typespecFile}.`
                    );
                    return true;
                }

                if (isDataPlane && IsReadmeUnderRoot.test(readmeFile)) {
                    context.logger.info(
                        `Exclude the readmeFile from SDK generation: ${readmeFile} because it is under data-plane root and is duplicated with typespec: ${typespecFile}.`
                    );
                    return true;
                }

                const typespecOutputFolder = getTypeSpecOutputFolder(
                    typespecFile,
                    context
                );
                typespecOutputFolderRegExp =
                    typespecOutputFolderRegExp ||
                    new RegExp(typespecOutputFolder, "ig");
                if (
                    isDataPlane &&
                    typespecOutputFolderRegExp.test(readmeFile)
                ) {
                    context.logger.info(
                        `Exclude the readmeFile from SDK generation: ${readmeFile} because it is duplicated with typespec: ${typespecFile}, typespecOutputFolder: ${typespecOutputFolder}.`
                    );
                    return true;
                }
            }
            return false;
        });
    });

    return filteredReadmeMdFiles;
}

/**
 * @param yamlContent
 * @returns {result: string | object | undefined | null, message: string}
 * special return
 * if the content is empty, return {result: null, message: string
 * if the file parse error, return {result: undefined, message: string
 */
export function parseYamlContent(
    yamlContent: string,
    path: string
): {
    result: string | object | undefined | null;
    message: string;
} {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let content: any = undefined;
    // if yaml file is not a valid yaml, catch error and return undefined
    try {
        content = YAML.parse(yamlContent);
    } catch (error) {
        console.error(
            `The file parsing failed in the ${path}. Details: ${error}`
        );
        return {
            result: content,
            message: `The file parsing failed in the ${path}. Details: ${error}`,
        };
    }

    // if yaml file is empty, run yaml.safeload success but get undefined
    // to identify whether it is empty return null to distinguish.
    if (!content) {
        console.info(
            `The file in the ${path} has been successfully parsed, but it is an empty file.`
        );
        return {
            result: null,
            message: `The file in the ${path} has been successfully parsed, but it is an empty file.`,
        };
    }

    return {
        result: content,
        message: "The file has been successfully parsed.",
    };
}

/**
 * Replace all of the instances of searchValue in value with the provided replaceValue.
 * @param {string | undefined} value The value to search and replace in.
 * @param {string} searchValue The value to search for in the value argument.
 * @param {string} replaceValue The value to replace searchValue with in the value argument.
 * @returns {string | undefined} The value where each instance of searchValue was replaced with replacedValue.
 */
export function replaceAll(value: string | undefined, searchValue: string, replaceValue: string): string | undefined {
  return !value || !searchValue ? value : value.split(searchValue).join(replaceValue || "");
}

/**
 * Extract and format the prefix from tspConfigPath or readmePath.
 * @param {string | undefined} tspConfigPath The tspConfigPath to extract the prefix from.
 * @param {string | undefined} readmePath The readmePath to extract the prefix from.
 * @returns {string} The formatted prefix.
 */
export function extractPathFromSpecConfig(tspConfigPath: string | undefined, readmePath: string | undefined): string {
  let prefix = '';
  if (tspConfigPath) {
    const match = tspConfigPath.match(/specification\/(.+)\/tspconfig\.yaml$/);
    if (match) {
      const segments = match[1].split('/');
      prefix = segments.join('-').toLowerCase().replace(/\./g, '-');
    }
  } else if (readmePath) {
    const match = readmePath.match(/specification\/(.+?)\/readme\.md$/i);
    if (match) {
      const segments = match[1].split('/');
      prefix = segments.join('-').toLowerCase().replace(/\./g, '-');
    }
  }
  return prefix;
}
