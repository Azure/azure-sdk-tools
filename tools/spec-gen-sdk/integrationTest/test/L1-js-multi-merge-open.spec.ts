import { createPRWithPatch, getPRListInfo, getLoggerTextForAssert, getTestGithubClient, repoOwner } from '../utils';
import { fixtures } from '../fixtures';
import { prepareJsSDKTest } from './lang-common';

describe('In open PR, SDK Automation', () => {
  it('should run in open pr after one merged pr', async () => {
    const { specRepoName, sdkRepoName, logger, launchSDKAutomation, patchJsLog } = await prepareJsSDKTest(
      'multi-merge-open'
    );
    const github = await getTestGithubClient();

    // Simulate spec PR merged and integration branch updated
    let specPRNumber = await createPRWithPatch(specRepoName, fixtures.specTest.patch0_AddService, 'patch0');
    await github.pulls.merge({ owner: repoOwner, repo: specRepoName, pull_number: specPRNumber });
    await createPRWithPatch(sdkRepoName, fixtures.sdkJs.patch0_AddServiceGen, 'sdkAutomation/@azure_test-service');

    // New open PR
    specPRNumber = await createPRWithPatch(specRepoName, fixtures.specTest.patch1_Add_02_01, 'patch1', {
      fetchBase: true
    });

    // Launch SDK Automation
    await launchSDKAutomation(specPRNumber);

    // Assert
    expect(await getPRListInfo(sdkRepoName)).toMatchSnapshot();
    expect(patchJsLog(await getLoggerTextForAssert(logger, specRepoName, specPRNumber, sdkRepoName))).toMatchSnapshot();
  });
});
