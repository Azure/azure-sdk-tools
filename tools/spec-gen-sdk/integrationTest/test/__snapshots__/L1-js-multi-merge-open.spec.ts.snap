// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`In open PR, SDK Automation should run in open pr after one merged pr 1`] = `
Array [
  Object {
    "base": "master",
    "head": "sdkAutomation/@azure_test-service",
    "number": 1,
    "state": "open",
    "title": "sdkAutomation/@azure_test-service",
  },
  Object {
    "base": "master",
    "head": "sdkAuto/2/@azure_test-service",
    "number": 2,
    "state": "open",
    "title": "[AutoPR @azure/test-service] patch1",
  },
]
`;

exports[`In open PR, SDK Automation should run in open pr after one merged pr 2`] = `
"github	Get specificationRepositoryConfiguration.json from <org>/<spec_repo>
info	mainRepository: <org>/<sdk_repo>
info	mainBranch: master
info	integrationRepository: <org>/<sdk_repo>
info	integrationBranchPrefix: sdkAutomation
info	secondaryRepository: <org>/<sdk_repo>
info	secondaryBranch: master
github	Get swagger_to_sdk_config.json from <org>/<sdk_repo>
git	Set remote main on <org>/<spec_repo> with auth
git	Delete branch refs/heads/master
git	Delete branch refs/heads/patch0
git	Delete branch refs/heads/patch1
git	Fetch pull/2/merge on <org>/<spec_repo>
git	Set remote main on <org>/<sdk_repo> with auth
git	Set remote secondary on <org>/<sdk_repo> with auth
git	Set remote integration on <org>/<sdk_repo> with auth
git	Delete branch refs/heads/master
git	Delete branch refs/heads/sdkAutomation/@azure_test-service
git	Fetch <org>/<sdk_repo> to main
git	Checkout main
git	Checkout secondary from main
warn	Skip initScript due to not configured
info	5 changed files in spec PR:
info	A	specification/testservice/resource-manager/Microsoft.TestService/stable/2020-02-01/TestService.json
info	A	specification/testservice/resource-manager/Microsoft.TestService/stable/2020-02-01/examples/TestGet.json
info	A	specification/testservice/resource-manager/Microsoft.TestService/stable/2020-02-01/scenarios/shouldbeexcluded.json
info	M	specification/testservice/resource-manager/readme.md
info	M	specification/testservice/resource-manager/readme.python.md
info	Related readme.md list:
info		specification/testservice/resource-manager/readme.md
info	Handle the following readme.md:
info		specification/testservice/resource-manager/readme.md
git	Checkout branch sdkGen
section	Call legacy generate
command	autorest --version=V2 --typescript --license-header=MICROSOFT_MIT_NO_VERSION --use=@microsoft.azure/autorest.typescript@4.4.4 --typescript-sdks-folder=<work_dir>/<sdk_repo> ../<spec_repo>/specification/testservice/resource-manager/readme.md
info	Config: {\\"stdout\\":{\\"showInComment\\":{}},\\"stderr\\":{\\"showInComment\\":true,\\"scriptWarning\\":{}},\\"exitCode\\":{\\"showInComment\\":true,\\"result\\":\\"error\\"}}
section	Commit generate result
git	Add * in SDK repo
info	13 changed files after SDK generate:
info	A	sdk/testservice/arm-testservice/LICENSE.txt
info	A	sdk/testservice/arm-testservice/README.md
info	A	sdk/testservice/arm-testservice/package.json
info	A	sdk/testservice/arm-testservice/rollup.config.js
info	A	sdk/testservice/arm-testservice/src/models/index.ts
info	A	sdk/testservice/arm-testservice/src/models/mappers.ts
info	A	sdk/testservice/arm-testservice/src/models/parameters.ts
info	A	sdk/testservice/arm-testservice/src/models/testMappers.ts
info	A	sdk/testservice/arm-testservice/src/operations/index.ts
info	A	sdk/testservice/arm-testservice/src/operations/test.ts
info	A	sdk/testservice/arm-testservice/src/testServiceClient.ts
info	A	sdk/testservice/arm-testservice/src/testServiceClientContext.ts
info	A	sdk/testservice/arm-testservice/tsconfig.json
git	Commit all the changes
section	Detect changed packages
info	Package from changed file search: /^package\\\\.json$/
info	1 packages found after generation:
info		sdk/testservice/arm-testservice
git	Set remote main on <org>/<sdk_repo> with auth
git	Set remote secondary on <org>/<sdk_repo> with auth
git	Set remote integration on <org>/<sdk_repo> with auth
section	Handle package @azure/test-service
info	Package log to https://sdkautotest/dl/<org>/2/azure-sdk-for-js/@azure_test-service/logs.txt
info	baseBranch [master] baseRepo [<org>/<sdk_repo>] baseRemote [integration]
info	targetBranch [sdkAuto/2/@azure_test-service] targetRepo [<org>/<sdk_repo>] baseRemote [integration]
info	Skip sync baseBranch due to same remote same branch
git	Create targetBranch sdkAuto/2/@azure_test-service
git	Checkout sdk folders from sdkGen and commit: sdk/testservice/arm-testservice
git	Push sdkAuto/2/@azure_test-service to <org>/<sdk_repo>
section	Call legacy build for azure-sdk-for-js
command	npm pack --silent 
info	Config: {\\"exitCode\\":{\\"showInComment\\":true,\\"result\\":\\"error\\"},\\"stderr\\":{\\"showInComment\\":true},\\"logPrefix\\":\\"npmPack\\"}
info	changelogScript is not configured
info	Use legacy artifact search option
info	Search artifact with: /\\\\.tgz$/
info	1 artifact found:
info		sdk/testservice/arm-testservice/azure-test-service-1.0.0.tgz
info	Upload 1 artifact:
info		azure-test-service-1.0.0.tgz
info	Legacy InstallInstruction
github	Get PR in <org>/<sdk_repo> from <org>:sdkAuto/2/@azure_test-service to master
github	Create new PR
github	PR created at https://github.com/<org>/<sdk_repo>/pull/2
section	Report status
info	Main status [succeeded] hasBreakingChange [false]
info	Rendered commentSubTitle: 
info	Rendered commentBody: <ul>
  <li>
    <div>
      <code>️✔️</code>Succeeded Generate from
      <commit_sha>. SDK Automation <version>
    </div>
    <pre>
warn	Skip initScript due to not configured
command	autorest --version&#x3D;V2 --typescript --license-header&#x3D;MICROSOFT_MIT_NO_VERSION --use&#x3D;@microsoft.azure/autorest.typescript@4.4.4 --typescript-sdks-folder&#x3D;<work_dir>/<sdk_repo> ../<spec_repo>/specification/testservice/resource-manager/readme.md</pre
    >
  </li>
  <li>
    <div>
      <span
        ><code>️✔️</code><b>@azure/test-service</b> [<a
          href=\\"https://sdkautotest/dl/<org>/2/azure-sdk-for-js/@azure_test-service/logs.txt\\"
          >View full logs</a
        >]&nbsp; [<a
          href=\\"https://github.com/<org>/<sdk_repo>/pull/2\\"
          >Preview SDK Changes</a
        >]
      </span>
    </div>
    <ul>
      <li>
        <a
          href=\\"https://sdkautotest/dl/<org>/2/azure-sdk-for-js/@azure_test-service/azure-test-service-1.0.0.tgz\\"
          >azure-test-service-1.0.0.tgz</a
        >
      </li>
    </ul>
    <pre>
    >
  </li>
</ul>

info	Writing unified pipeline message to pipe.log"
`;
